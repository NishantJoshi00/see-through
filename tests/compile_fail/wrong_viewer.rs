use see_through::{see_derive::Look, Look, See};

#[derive(Look)]
struct Card<T> {
    name: String,
    number: T,
}

#[derive(Look)]
struct Shoe<T> {
    brand: T,
    number: i32,
}


fn change_i32_number<T>(mut var: T, by: i32) -> T
where
    T: Look<see_t::number, Inner = i32>,
{
    var[see_t::number] += by;
    var
}


fn main() {
    
    let shoe = Shoe {
        brand: "my",
        number: 2,
    };
    let card = Card {
        name: "dev".to_string(),
        number: "2",
    };

    let card = change_i32_number(card, 2); // <-- This won't work
    let shoe = change_i32_number(shoe, 2);
}


see_derive::auto_load!();
