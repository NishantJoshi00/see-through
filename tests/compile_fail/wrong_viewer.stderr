error[E0271]: type mismatch resolving `<Card<&str> as See<NUMBER>>::Inner == i32`
  --> tests/compile_fail/wrong_viewer.rs:36:34
   |
36 |     let card = change_i32_number(card, 2); // <-- This won't work
   |                ----------------- ^^^^ expected `i32`, found `&str`
   |                |
   |                required by a bound introduced by this call
   |
note: required by a bound in `change_i32_number`
  --> tests/compile_fail/wrong_viewer.rs:18:28
   |
16 | fn change_i32_number<T>(mut var: T, by: i32) -> T
   |    ----------------- required by a bound in this
17 | where
18 |     T: Look<see_t::NUMBER, Inner = i32>,
   |                            ^^^^^^^^^^^ required by this bound in `change_i32_number`
